src/box/relay.cc

```diff
@@ -1050,11 +1050,16 @@ relay_subscribe(struct replica *replica, struct iostream *io, uint64_t sync,
 	struct relay *relay = replica->relay;
 	assert(relay->state != RELAY_FOLLOW);
 	/*
-	 * Register the replica with the garbage collector
-	 * unless it has already been registered by initial
-	 * join.
+	 * Register the replica with the garbage collector.
+	 * In case some of the replica's WAL files were deleted, it might
+	 * subscribe with a smaller vclock than the master remembers, so
+	 * recreate the gc consumer unconditionally to make sure it holds
+	 * the correct vclock.
 	 */
-	if (replica->gc == NULL && !replica->anon) {
+	if (!replica->anon) {
+		if (replica->gc != NULL) {
+			gc_consumer_unregister(replica->gc);
+		}
```

Comment:
We've changed the behaviour here a bit: note the `gc_delay_unref()` a couple of lines below.
Previously in ran only when the replica came for the first time (`replica->gc == NULL`). And now it runs on every replica subscribe. Even if the replica resubscribes multiple times.

This is not good, one replica might trick the master into thinking that enough consumers are registered and make it delete still needed xlogs.

Let's return the old behaviour:
```c
bool had_gc = false;
if (replica->gc != NULL) {
    gc_consumer_unregister(replica->gc);
    had_gc = true;
}
```

And below:
```c
if (!had_gc)
    gc_delay_unref();
```

Reply:
Done